### ISSUE TITLE
torch.compile doesn't work with hasattr

### ISSUE THREAD

**Issue by SunMarc** [pytorch/pytorch#134577]
### 🐛 Describe the bug

torch.compile doesn't work when used with `hasattr`. It only works when when the condition is True. This is needed in transformers/accelerate in order to use compiled model on multi-gpu. cc @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @rec @vmoens @anijain2305 
Also, see related issue: https://github.com/huggingface/accelerate/issues/3000

### Minified repro

```py 
import torch
from tensordict import TensorDict

@torch.compile(fullgraph=True)
def func(x):
    if hasattr(x, "to"):
        return x.to("cpu")
    return x
func(dict(a=torch.randn(3)))
```

Thanks @vmoens for the example

### Versions

[pip3] mypy==1.5.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.4
[pip3] onnx==1.15.0
[pip3] onnxconverter-common==1.13.0
[pip3] pytorch-triton==3.0.0+dedb7bdf33
[pip3] q-galore-torch==1.0
[pip3] tf2onnx==1.16.1
[pip3] torch==2.5.0.dev20240729+cu121
[pip3] torchao==0.4.0
[pip3] torchao-nightly==2024.7.29+cu121
[pip3] torchaudio==2.4.0.dev20240729+cu121
[pip3] torchpippy==0.1.1+e9e2d5f
[pip3] torchvision==0.20.0.dev20240729+cu121
[pip3] triton==3.0.0

**Comment by anijain2305** [comment]
https://github.com/pytorch/pytorch/pull/134590 Fixes at least this error. We might have missing support for many other types of variable tracker.

**Comment by vmoens** [comment]
@anijain2305 can maybe confirm but if I run

```python
import torch
from tensordict import TensorDict

@torch.compile(fullgraph=True)
def func(x):
    if hasattr(x, "to"):
        return x.to("cpu")
    return x
func(torch.randn(3))
```
on nightly, or 
```python
import torch
from tensordict import TensorDict

@torch.compile(fullgraph=True)
def func(x):
    if hasattr(x, "hocuspocus"):
        return x.to("cpu")
    return x
func(torch.randn(3))
```
both tests pass which I think indicates that the problem is solved

Not 100% sure it made it through the rc branch though

**Comment by zyxiyy** [comment]
Unsupported: hasattr SkipFunctionVariable to

from user code:
   File "/tmp/ipykernel_1957076/1822748636.py", line 7, in decode_one_tokens
    logits = model(
  File "/home/bcds/.conda/envs/llm/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bcds/.conda/envs/llm/lib/python3.9/site-packages/accelerate/hooks.py", line 165, in new_forward
    args, kwargs = module._hf_hook.pre_forward(module, *args, **kwargs)
  File "/home/bcds/.conda/envs/llm/lib/python3.9/site-packages/accelerate/hooks.py", line 364, in pre_forward
    return send_to_device(args, self.execution_device), send_to_device(
  File "/home/bcds/.conda/envs/llm/lib/python3.9/site-packages/accelerate/utils/operations.py", line 184, in send_to_device
    {
  File "/home/bcds/.conda/envs/llm/lib/python3.9/site-packages/accelerate/utils/operations.py", line 185, in <dictcomp>
    k: t if k in skip_keys else send_to_device(t, device, non_blocking=non_blocking, skip_keys=skip_keys)
  File "/home/bcds/.conda/envs/llm/lib/python3.9/site-packages/accelerate/utils/operations.py", line 149, in send_to_device
    if is_torch_tensor(tensor) or hasattr(tensor, "to"):
    i have this problem when i compile mixtral with mutigpu.
